R/03_additionalLinters.R

```diff
@@ -21,85 +21,207 @@
-  pastedText <- lapply(1:(length(source_file$file_lines) - 1),
-                       function(x) paste(source_file$file_lines[x],
-                                         source_file$file_lines[x + 1]))
-
-  idsMult <- grep(pattern, pastedText)
-
-  # Remove double-counted problems
-  ids <- c(idsWithin, setdiff(idsMult, c(idsWithin - 1))) %>% unique
-
-  lapply(ids, function(id) {
-    Lint(filename = source_file$filename,
-         line_number = id,
-         column_number = 1L,
-         type = "style",
-         message = "Arguments without default value should be listed before
-         arguments with default value.")
-  })},
-  "args_no_default_first_linter")
-}
-
-
 #' @describeIn INWTLinters Are there double whitespaces?
 #' @export
 double_space_linter <- function() {
-  Linter(function(source_file) {
-
-  ids <- grep("([^#' ]+.*[^ ]+ {2,}[^( +#$)])|(^# \\w {2,})",
-              source_file$file_lines)
-  # At least one character which is neither # nor '
-  # Any characters
-  # At least one character which is not space
-  # At least two spaces
-  # ...not directly followed by some spaces and a hash
-  # OR: Starting with #, then space, then word character, then at least two
-  # spaces
-  lapply(ids, function(id) {
-    Lint(filename = source_file$filename,
-         line_number = id,
-         column_number = 1L,
-         type = "style",
-         message = "Double whitespace.")
-  })},
-  "double_space_linter")
+  # X-Path logic to find the deepest node with in-text double spaces. Either
+  # COMMENT or STR_CONST.
+  comment_string_xpath <- "//descendant::*[contains(., '  ') and not(descendant::*[contains(., '  ')])]"
+  # Same logic for non-in-text double spaces (in remaining script content).
+  code_xpath <- "//descendant::*[not(descendant::*) and @end and not(ancestor-or-self::COMMENT)]"
+  Linter(function(source_expression) {
+    if (!is_lint_level(source_expression, level = "file")) {
+      return(list())
+    }
+    # Set the correct format of the xml-parsed script.
+    xml <- source_expression$full_xml_parsed_content
+
+    ### comment lints ###
+    comment_string_nodes <- xml_find_all(xml, comment_string_xpath)
+    # Inputs the node list above and collects the locations of the in-text
+    # double spaces. For each node, a df containing the start and end points is
+    # created.
+    filter_double_space_nodes <- function(nodes) {
+      valid_nodes <- list()
+
+      for (i in seq_along(nodes)) {
+        text <- xml_text(nodes[[i]])
+        double_spaces <- str_locate_all(text, "\\s{2,}")[[1]]
+
+        if (nrow(double_spaces) > 0) {
+          info_df <- as.data.frame(double_spaces)
+          starts <- info_df$start
+          ends <- info_df$end
+
+          has_invalid_match <- FALSE
+
+          for (j in seq_along(starts)) {
+            start_pos <- starts[j]
+            end_pos <- ends[j]
+            char_before <- if (start_pos > 1) substr(text, start_pos - 1, start_pos - 1) else ""
```

Comment:
```suggestion
            char_before <- ifelse((start_pos > 1), substr(text, start_pos - 1, start_pos - 1), "")
```
Die Zeile darunter analog